File: client_android/MiniServiceDiscover/src/com/miniservice/discover/client/Protocal.java
package com.miniservice.discover.client;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;
import android.util.Log;

public class Protocal {
    private final static String TAG = "Protocal";
    public final static int PROTOCAL_LEN = 1024;
    public final static short INFO_HEAD = (short)0xFFEE;
    public final static short INFO_TYPE_REGISTER_SERVICE   = 0;
    public final static short INFO_TYPE_UNREGISTER_SERVICE = 1;
    public final static short INFO_TYPE_INQUIRY            = 2;
    public final static short INFO_TYPE_ACTION_SUCC        = 3;
    public final static short INFO_TYPE_ACTION_FAIL        = 4;
    public final static short INFO_TYPE_ACTION_INVALID     = -1;
   

    public static class protocalHead{
        public short head;
        public short action;
        public short len;
        public short checksum;
        
        private short getShort(byte[] data, int offset){
            return (short) (((data[offset + 1] << 8) | data[offset + 0] & 0xff));  
        }
        
        public protocalHead() {}
        
        public protocalHead(byte[] data) {
            head =  getShort(data, 0);
            action =  getShort(data, 2);
            len =  getShort(data, 4);
            checksum = getShort(data, 6);
        }
        
        public int getSize(){
            return Short.SIZE/8*4;
        }
        
        public byte[] getBytes(){
            byte[] data = new byte[getSize()];
            int i = 0;
            
            data[i++] = (byte)(head&0xFF);
            data[i++] = (byte)((head>>8)&0xFF);
            data[i++] = (byte)(action&0xFF);
            data[i++] = (byte)((action>>8)&0xFF);
            data[i++] = (byte)(len&0xFF);
            data[i++] = (byte)((len>>8)&0xFF);
            data[i++] = (byte)(checksum&0xFF);
            data[i++] = (byte)((checksum>>8)&0xFF);
            return data;
        }
    };
    
    private short getChecksumByOffset(byte[] data, int offset, int len){
        short sum = 0;
        //Log.i(TAG, new String(data, offset, len));
        for(int i=0; i<len; i++){
            sum +=data[offset+i];
        }
        return sum;
    }
    
    private short getChecksum(byte[] data){
        return getChecksumByOffset(data, 0, data.length);
    }
    
    public byte[] getInquiryPkt(String serviceName){
        byte[] data = new byte[PROTOCAL_LEN];
        
        protocalHead head = new protocalHead();
        head.head = INFO_HEAD;
        head.action = INFO_TYPE_INQUIRY;
        
        Map map = new HashMap();
        map.put("device_name", "");
        map.put("service_name", serviceName);
        map.put("ip", "");
        map.put("port", 0);
        
        JSONObject json = new JSONObject(map);
        byte[] jsonData = json.toString().getBytes();
        
        head.len = (short)jsonData.length;
        head.checksum = getChecksum(jsonData);
        
        //Log.i(TAG, "getInquiryPkt json:"+jsonData.toString()+" len="+jsonData.toString().length()+" checksum:"+head.checksum);
        
        System.arraycopy(head.getBytes(), 0, data, 0,  head.getSize());
        System.arraycopy(jsonData,        0, data, head.getSize(),  jsonData.length);
        return data;
    }
    
    
    public int decodeAction(byte[] data) {
        protocalHead  head = new protocalHead(data);
        //Log.i(TAG, "head"+head.head);
        //Log.i(TAG, "action"+head.action);
        //Log.i(TAG, "len"+head.len);
        //Log.i(TAG, "checksum"+head.checksum);
        short checksum = getChecksumByOffset(data, head.getSize(), head.len);
        if(checksum == head.checksum){
            return head.action;
        } else {
            Log.e(TAG, "checksum fail, cal="+checksum+" rec="+head.checksum);
            return INFO_TYPE_ACTION_INVALID;
        }
    }
    
    public ServiceInfo getServiceInfo(byte[] data){
        ServiceInfo info = new ServiceInfo();
        protocalHead  head = new protocalHead(data);

        try {
            String json = new String(data, head.getSize(), head.len, "UTF-8");
            //Log.i(TAG, "getServiceInfo json:"+json);
            try{
                JSONObject jsonObj = new JSONObject(json);
                info.dev = jsonObj.getString("device_name");
                info.name = jsonObj.getString("service_name");
                info.ip = jsonObj.getString("ip");
                info.port = jsonObj.getInt("port");
            }catch(JSONException e) {  
                e.printStackTrace();  
            }  
        } catch (UnsupportedEncodingException e1) {
            e1.printStackTrace();
        }
       
        return info;
    }
    
    public String getResult(byte[] data){
        protocalHead  head = new protocalHead(data);
        
        byte[] jsonData = new byte[head.len];
        System.arraycopy(data, head.getSize(), jsonData, 0,  head.len);
        
        try {
            String json = new String(data, head.getSize(), head.len, "UTF-8");
            try{
                JSONObject jsonObj = new JSONObject(json);
                return jsonObj.getString("reason");
            }catch(JSONException e) {  
                e.printStackTrace();  
            }  
        } catch (UnsupportedEncodingException e1) {
            e1.printStackTrace();
        }
        return "";
    }
}

File: c/provider/comm/protocal.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "protocal.h"
#include "cJSON.h"
#include "udp.h"



service_pkt *get_pkt()
{
    service_pkt *pkt = (service_pkt *)malloc(sizeof(service_pkt));
    bzero(pkt, sizeof(service_pkt));
    
    pkt->data = (uint8_t *)malloc(UDP_COMM_BUF_LEN);
    bzero(pkt->data, UDP_COMM_BUF_LEN);
    pkt->len = UDP_COMM_BUF_LEN;
    
    return pkt;
}

void free_pkt(service_pkt *pkt)
{
    free(pkt->data);
    free(pkt);
}

uint16_t get_checksum(char *data, uint16_t len)
{
    uint16_t checksum = 0;
    
    for(uint16_t i=0; i<len; i++){
        checksum += *data++;
    }
    
    return checksum;
}

service_pkt *get_register_pkt(char *device_name, char *service_name, char *ip,uint16_t port, char *info)
{
    service_pkt *pkt = get_pkt();
    char device_name_buf[INFO_MAX_LEN];
    char service_name_buf[INFO_MAX_LEN];
    char ip_buf[INFO_MAX_LEN];
    char info_buf[INFO_MAX_LEN];
    
    strncpy(device_name_buf, device_name, sizeof(device_name_buf));
    strncpy(service_name_buf, service_name, sizeof(service_name_buf));
    strncpy(ip_buf, ip, sizeof(ip_buf));
    strncpy(info_buf, info, sizeof(info_buf));
    
    //content
    service_head* head = (service_head*)pkt->data;
    head->head = INFO_HEAD;
    head->action = INFO_TYPE_REGISTER_SERVICE;

    cJSON *root = cJSON_CreateObject();  
    
    cJSON_AddStringToObject(root,"device_name",  device_name_buf);  
    cJSON_AddStringToObject(root,"service_name", service_name_buf);  
    cJSON_AddStringToObject(root,"ip",   ip_buf);  
    cJSON_AddStringToObject(root,"info", info_buf); 
    cJSON_AddNumberToObject(root,"port", port);  
    
    char *out = cJSON_Print(root);  
    //printf("%s\n",out);  
    
    head->info_len = strlen(out);
    memcpy(pkt->data+sizeof(service_head), out, head->info_len);
    
    head->checksum = get_checksum(out, head->info_len);
    
    cJSON_Delete(root);  
    free(out);
    
    return pkt;
}

service_pkt *get_unregister_pkt(char *device_name, char *service_name)
{
    service_pkt *pkt = get_pkt();
    char device_name_buf[INFO_MAX_LEN];
    char service_name_buf[INFO_MAX_LEN];
    
    strncpy(device_name_buf, device_name, sizeof(device_name_buf));
    strncpy(service_name_buf, service_name, sizeof(service_name_buf));
    
    //content
    service_head* head = (service_head*)pkt->data;
    head->head = INFO_HEAD;
    head->action = INFO_TYPE_UNREGISTER_SERVICE;

    cJSON *root = cJSON_CreateObject();  

    cJSON_AddStringToObject(root,"device_name", device_name_buf);  
    cJSON_AddStringToObject(root,"service_name", service_name_buf);  
    cJSON_AddStringToObject(root,"ip", "");  
    cJSON_AddStringToObject(root,"info", ""); 
    cJSON_AddNumberToObject(root,"port", 0);  
    
    char *out = cJSON_Print(root);  
    //printf("%s\n",out);  
    
    head->info_len = strlen(out);
    memcpy(pkt->data+sizeof(service_head), out, head->info_len);
    
    head->checksum = get_checksum(out, head->info_len);
    
    cJSON_Delete(root);  
    free(out);
    
    return pkt;
}

service_pkt *get_inquiry_pkt(char *service_name)
{
    service_pkt *pkt = get_pkt();
    char service_name_buf[INFO_MAX_LEN];
    
    strncpy(service_name_buf, service_name, sizeof(service_name_buf));
    
    //content
    service_head* head = (service_head*)pkt->data;
    head->head = INFO_HEAD;
    head->action = INFO_TYPE_INQUIRY;

    cJSON *root = cJSON_CreateObject();  

    cJSON_AddStringToObject(root,"device_name", "");  
    cJSON_AddStringToObject(root,"service_name", service_name_buf);  
    cJSON_AddStringToObject(root,"ip", "");  
    cJSON_AddStringToObject(root,"info", ""); 
    cJSON_AddNumberToObject(root,"port", 0);  
    
    char *out = cJSON_Print(root);  
    //printf("%s\n",out);  
    
    head->info_len = strlen(out);
    memcpy(pkt->data+sizeof(service_head), out, head->info_len);
    
    head->checksum = get_checksum(out, head->info_len);
    
    cJSON_Delete(root);  
    free(out);
    
    return pkt;
}

service_pkt *get_inquiry_result_pkt(service_info *info)
{
    service_pkt *pkt = get_pkt();
    char device_name_buf[INFO_MAX_LEN];
    char service_name_buf[INFO_MAX_LEN];
    char ip_buf[INFO_MAX_LEN];
    char info_buf[INFO_MAX_LEN];
    
    strncpy(device_name_buf, info->device_name, sizeof(device_name_buf));
    strncpy(service_name_buf, info->service_name, sizeof(service_name_buf));
    strncpy(ip_buf, info->ip, sizeof(ip_buf));
    strncpy(info_buf, info->info, sizeof(info_buf));
    
    //content
    service_head* head = (service_head*)pkt->data;
    head->head = INFO_HEAD;
    head->action = INFO_TYPE_ACTION_SUCC;

    cJSON *root = cJSON_CreateObject();  

    cJSON_AddStringToObject(root,"device_name",  device_name_buf);  
    cJSON_AddStringToObject(root,"service_name", service_name_buf);  
    cJSON_AddStringToObject(root,"ip",   ip_buf);  
    cJSON_AddStringToObject(root,"info", info_buf); 
    cJSON_AddNumberToObject(root,"port", info->port);  
    
    char *out = cJSON_Print(root);  
    //printf("%s\n",out);  
    
    head->info_len = strlen(out);
    memcpy(pkt->data+sizeof(service_head), out, head->info_len);
    
    head->checksum = get_checksum(out, head->info_len);
    
    cJSON_Delete(root);  
    free(out);
    
    return pkt;
}


service_pkt *get_result_pkt(int is_succ, char *reason)
{
    service_pkt *pkt = get_pkt();
    char reason_buf[INFO_MAX_LEN];
    
    strncpy(reason_buf, reason, sizeof(reason_buf));
    
    //content
    service_head* head = (service_head*)pkt->data;
    head->head = INFO_HEAD;
    head->action = is_succ?INFO_TYPE_ACTION_SUCC:INFO_TYPE_ACTION_FAIL;

    cJSON *root = cJSON_CreateObject();  

    cJSON_AddStringToObject(root,"reason", reason_buf);
    
    char *out = cJSON_Print(root);  
    //printf("%s\n",out);  
    
    head->info_len = strlen(out);
    memcpy(pkt->data+sizeof(service_head), out, head->info_len);
    
    head->checksum = get_checksum(out, head->info_len);
    
    cJSON_Delete(root);  
    free(out);
    
    return pkt;    
}


//>0: succ
int checksum_right(service_pkt *pkt)
{
    service_head* head = (service_head*)pkt->data;
    
    uint16_t sum = 0;
    uint8_t *data = pkt->data + sizeof(service_head);
    for(uint16_t i=0; i<head->info_len; i++){
        sum += *data++;
    }
    
    return sum == head->checksum;
}

ACTION_TYPE decode_action(service_pkt *pkt)
{
    if(checksum_right(pkt)){
        service_head* head = (service_head*)pkt->data;
        return head->action;
    } else {
        return INFO_TYPE_ACTION_INVALID;
    }
}

service_info* get_service_info(service_pkt *pkt)
{
    const char *json_data = (const char *)(pkt->data + sizeof(service_head));
    service_info* info;
    
        
    cJSON *json, *json_value;  
      
    json = cJSON_Parse(json_data);  
    if (!json)  
    {  
        printf("Error before: [%s]\n",cJSON_GetErrorPtr());  
        return NULL;
    }  
    else  
    {  
        info = (service_info *)malloc(sizeof(service_info));
        bzero(info, sizeof(service_info));

        json_value = cJSON_GetObjectItem( json , "device_name");  
        if(json_value && json_value->type == cJSON_String )  {  
            info->device_name = strdup(json_value->valuestring);
        } 
        json_value = cJSON_GetObjectItem( json , "service_name");  
        if(json_value && json_value->type == cJSON_String )  {  
            info->service_name = strdup(json_value->valuestring);
        } 
        json_value = cJSON_GetObjectItem( json , "ip");  
        if(json_value && json_value->type == cJSON_String )  {  
            info->ip = strdup(json_value->valuestring);
        }
        json_value = cJSON_GetObjectItem( json , "info");  
        if(json_value && json_value->type == cJSON_String )  {  
            info->info = strdup(json_value->valuestring);
        } 
        json_value = cJSON_GetObjectItem( json , "port");  
        if(json_value && json_value->type == cJSON_Number )  {  
            info->port = json_value->valueint;
        } 
        
        cJSON_Delete(json);  
        return info;
    }  

 

    return NULL;
}

void free_service_info(service_info *info)
{
    if(info->device_name)
       free(info->device_name);
    if(info->service_name)
       free(info->service_name);
    if(info->ip)
       free(info->ip);
    free(info);
}

result_info* get_result(service_pkt *pkt)
{
    service_head* head = (service_head*)pkt->data;
    const char *json_data = (const char *)(pkt->data + sizeof(service_head));
    result_info* res;
    
    if(head->action == INFO_TYPE_ACTION_FAIL){
        
        cJSON *json, *json_value;  
          
        json = cJSON_Parse(json_data);  
        if (!json)  
        {  
            printf("Error before: [%s]\n",cJSON_GetErrorPtr());  
            return NULL;
        }  
        else  
        {  
            res = (result_info *)malloc(sizeof(result_info));
            bzero(res, sizeof(result_info));
    
            json_value = cJSON_GetObjectItem( json , "reason");  
            if(json_value && json_value->type == cJSON_String )  {  
                res->reason = strdup(json_value->valuestring);
            } 
            
            cJSON_Delete(json);  
            return res;
        }  
    
    }
    
    return NULL;
}

void free_result(result_info *res)
{
    if(res->reason)
       free(res->reason);
    free(res);
}
